src/JustEat.StatsD/SimpleObjectPool.cs

```diff
@@ -1,56 +1,67 @@
 using System;
 using System.Collections.Concurrent;
-using System.Diagnostics.CodeAnalysis;
 
 namespace JustEat.StatsD
 {
-    /// <summary>	A class that provides simple thread-safe object pooling semantics.  </summary>
-    public sealed class SimpleObjectPool<T>
-        where T : class
+    /// <summary>
+    /// A class that provides simple thread-safe object pooling semantics.
+    /// </summary>
+    public sealed class SimpleObjectPool<T> where T : class
     {
+        private readonly Func<SimpleObjectPool<T>, T> _constructor;
         private readonly ConcurrentBag<T> _pool;
 
         /// <summary>	Constructor that populates a pool with the given number of items. </summary>
-        /// <exception cref="ArgumentNullException">	Thrown when the constructor is null. </exception>
-        /// <exception cref="ArgumentException">		Thrown when the constructor produces null objects. </exception>
-        /// <param name="capacity">   	The capacity. </param>
-        /// <param name="constructor">	The factory method used to create new instances of the object to populate the pool. </param>
-        [SuppressMessage("Microsoft.Design", "CA1006:DoNotNestGenericTypesInMemberSignatures", Justification = "Necessary to nest generics to support passing a factory method")]
-        public SimpleObjectPool(int capacity, Func<SimpleObjectPool<T>, T> constructor)
+        /// <exception cref="ArgumentNullException"> Thrown when the constructor is null. </exception>
+        /// <param name="initialSize"> Number of items in the pool at start </param>
+        /// <param name="constructor"> The factory method used to create new instances of the object to populate the pool. </param>
+        public SimpleObjectPool(int initialSize, Func<SimpleObjectPool<T>, T> constructor)
         {
-            if (constructor == null)
-            {
-                throw new ArgumentNullException(nameof(constructor));
-            }
-
+            _constructor = constructor ?? throw new ArgumentNullException(nameof(constructor));
             _pool = new ConcurrentBag<T>();
-            for (var i = 0; i < capacity; ++i)
+            PrePopulate(initialSize);
+        }
+
+        private void PrePopulate(int size)
+        {
+            while (Count < size)
             {
-                var instance = constructor(this);
+                var instance = _constructor(this);
                 if (instance == null)
                 {
-                    throw new ArgumentException("constructor produced null object", "constructor");
+                    throw new InvalidOperationException("constructor produced null object");
```

Comment:
Whether it's worth actually doing anything about to revert is up for debate, but based on the standards Microsoft adhere to, but I've just noticed that this is a breaking change.